logstash-core/src/test/java/org/logstash/secret/store/SecretStoreFactoryTest.java

```diff
@@ -0,0 +1,194 @@
+package org.logstash.secret.store;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+import org.logstash.secret.SecretIdentifier;
+import org.logstash.secret.store.backend.JavaKeyStore;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.logstash.secret.store.SecretStoreFactory.*;
+
+/**
+ * Unit tests for {@link SecretStoreFactory}
+ */
+public class SecretStoreFactoryTest {
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    //exact copy from https://stackoverflow.com/questions/318239/how-do-i-set-environment-variables-from-java
+    //thanks @pushy and @Edward Campbell !
+    @SuppressWarnings("unchecked")
+    private static void setEnv(Map<String, String> newenv) throws Exception {
+        try {
+            Class<?> processEnvironmentClass = Class.forName("java.lang.ProcessEnvironment");
+            Field theEnvironmentField = processEnvironmentClass.getDeclaredField("theEnvironment");
+            theEnvironmentField.setAccessible(true);
+            Map<String, String> env = (Map<String, String>) theEnvironmentField.get(null);
+            env.putAll(newenv);
+            Field theCaseInsensitiveEnvironmentField = processEnvironmentClass.getDeclaredField("theCaseInsensitiveEnvironment");
+            theCaseInsensitiveEnvironmentField.setAccessible(true);
+            Map<String, String> cienv = (Map<String, String>) theCaseInsensitiveEnvironmentField.get(null);
+            cienv.putAll(newenv);
+        } catch (NoSuchFieldException e) {
+            Class[] classes = Collections.class.getDeclaredClasses();
+            Map<String, String> env = System.getenv();
+            for (Class cl : classes) {
+                if ("java.util.Collections$UnmodifiableMap".equals(cl.getName())) {
+                    Field field = cl.getDeclaredField("m");
+                    field.setAccessible(true);
+                    Object obj = field.get(env);
+                    Map<String, String> map = (Map<String, String>) obj;
+                    map.clear();
+                    map.putAll(newenv);
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testAlternativeImplementation(){
+        String beforeSystem = System.getProperty(SecretStore.class.getCanonicalName());
+        try {
+            System.setProperty(SecretStore.class.getCanonicalName(), "org.logstash.secret.store.SecretStoreFactoryTest$MemoryStore");
```

Comment:
My instinct for java properties is that we should use `logstash.yml` settings for this. Can you tell me a story about why this is a java system property and not in Logstash settings?

Reply:
This has been replaced with a value from `logstash.yml` and will be part of a future review. _I sometimes use system properties as placeholders while building out functionality`

Reply:
It's hard to verify this claim, and it is likely I will have forgotten to check for it in a future review. It is hard for me to carry context of reviews across multiple PRs (we don't' do this for any other PRs?), so while I trust you, I cannot verify your claim that it will be part of a future review because I have little-or-no means to track the deferred work in this PR in a future PR.

Should I file blocker issues about all of these deferred things?

Reply:
Here is the (future) code that will read read from the settings:

https://github.com/jakelandis/logstash/blob/8353-working/logstash-core/lib/logstash/util/secretstore.rb#L15
and
https://github.com/jakelandis/logstash/blob/8353-working/logstash-core/src/main/java/org/logstash/secret/store/SecretStoreFactory.java#L70